casper/src/main/scala/coop/rchain/casper/LastFinalizedBlockCalculator.scala

```diff
@@ -0,0 +1,74 @@
+package coop.rchain.casper
+
+import cats.Monad
+import cats.effect.{Concurrent, Sync}
+import cats.implicits._
+import coop.rchain.blockstorage.{BlockDagRepresentation, BlockDagStorage, BlockStore}
+import coop.rchain.casper.util.ProtoUtil
+import coop.rchain.shared.{Cell, Log}
+import coop.rchain.catscontrib.ListContrib
+
+object LastFinalizedBlockCalculator {
+  // TODO: Extract hardcoded fault tolerance threshold
+  private val faultToleranceThreshold = 0f
+
+  def run[F[_]: Monad: Log: Sync: Concurrent: BlockStore: BlockDagStorage: SafetyOracle](
```

Comment:
Is this rather complex method tested anywhere (possibly on randomly generated input)?

Reply:
Basically any call to https://github.com/rchain/rchain/pull/2200/files/2c47336efce2a6afe0cd6b56abf7617a642bec36#diff-de23b4200eaecb35ba261ac3caa4d112R202 tests this

For example,

https://github.com/rchain/rchain/blob/0115d5854bcd5e31402b920a4f820b45ebaf0f1f/casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala#L1676

https://github.com/rchain/rchain/blob/0115d5854bcd5e31402b920a4f820b45ebaf0f1f/casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala#L1687

https://github.com/rchain/rchain/blob/0115d5854bcd5e31402b920a4f820b45ebaf0f1f/casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala#L1698

https://github.com/rchain/rchain/blob/0115d5854bcd5e31402b920a4f820b45ebaf0f1f/casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala#L1708

How would a randomly generated block dag testing work here?

Reply:
For starters you could generate random DAGs for which it is trivial to determine which blocks should be finalized and check if the algorithm properly marks them as finalized

Reply:
If it is a random DAG, then it isn't trivial to determine which blocks should be finalized - only the genesis block can always be trivially finalized.

Reply:
Ok